apis/projectcontour/v1alpha1/contourconfig.go

```diff
@@ -21,7 +21,519 @@ import (
 // ContourConfigurationSpec represents a configuration of a Contour controller.
 // It contains most of all the options that can be customized, the
 // other remaining options being command line flags.
+// +kubebuilder:default=xdsServer: {type: "contour", address: "0.0.0.0", port: 8001, tls: { caFile: "/certs/ca.crt", certFile: "/certs/tls.crt", keyFile: "/certs/tls.key", insecure: false}}, debug: {logLevel: "info", kubernetesLogLevel: 0}, health: {address: "0.0.0.0", port: 8000}, envoy: {listener: {useProxyProtocol: false, disableAllowChunkedLength: false, connectionBalancer: "", tls: { minimumProtocolVersion: "1.2", cipherSuites: "[ECDHE-ECDSA-AES128-GCM-SHA256|ECDHE-ECDSA-CHACHA20-POLY1305]";"[ECDHE-RSA-AES128-GCM-SHA256|ECDHE-RSA-CHACHA20-POLY1305]";"ECDHE-ECDSA-AES256-GCM-SHA384";"ECDHE-RSA-AES256-GCM-SHA384" }}, service: {name: "envoy", namespace: "projectcontour"}, http: {address: "0.0.0.0", port: 8080, accessLog: "/dev/stdout"}, https: {address: "0.0.0.0", port: 8443, accessLog: "/dev/stdout"}, metrics: {address: "0.0.0.0", port: 8002}, logging: { accessLogFormat: "envoy"}, defaultHTTPVersions: "http/1.1";"http/2", cluster: {dnsLookupFamily: "auto"}, network: { adminPort: 9001}}, httpproxy: {disablePermitInsecure: false}, leaderElection: {leaseDuration: "15s", renewDeadline: "10s", retryPeriod: "2s", disableLeaderElection: false, configmap: {name: "leader-elect", namespace: "projectcontour"}}, enableExternalNameService: false, metrics: {address: "0.0.0.0", port: 8000}
```

Comment:
One option to consider here would be leaving off this aggregated default for the `spec` field itself. Since `spec` is a required field, the UX would be that a user would create a `ContourConfig` like the following:

```yaml
kind: ContourConfiguration
apiVersion: projectcontour.io/v1alpha1
metadata:
  name: contour-config
  namespace: default
spec: {}
```

Then they'll still get all the defaults for the subfields of `spec`:

```yaml
apiVersion: projectcontour.io/v1alpha1
kind: ContourConfiguration
metadata:
  annotations:
    kubectl.kubernetes.io/last-applied-configuration: |
      {"apiVersion":"projectcontour.io/v1alpha1","kind":"ContourConfiguration","metadata":{"annotations":{},"name":"contour-config","namespace":"default"},"spec":{}}
  creationTimestamp: "2021-09-17T16:18:54Z"
  generation: 1
  name: contour-config
  namespace: default
  resourceVersion: "175456"
  uid: 907940dd-39c7-4fe5-ae14-001061dcb0aa
spec:
  debug:
    kubernetesLogLevel: 0
    logLevel: info
  enableExternalNameService: false
  envoy:
    cluster:
      dnsLookupFamily: auto
    defaultHTTPVersions:
    - http/1.1
    - http/2
    http:
      accessLog: /dev/stdout
      address: 0.0.0.0
      port: 8080
    https:
      accessLog: /dev/stdout
      address: 0.0.0.0
      port: 8443
    listener:
      connectionBalancer: ""
      disableAllowChunkedLength: false
      tls:
        cipherSuites:
        - '[ECDHE-ECDSA-AES128-GCM-SHA256|ECDHE-ECDSA-CHACHA20-POLY1305]'
        - '[ECDHE-RSA-AES128-GCM-SHA256|ECDHE-RSA-CHACHA20-POLY1305]'
        - ECDHE-ECDSA-AES256-GCM-SHA384
        - ECDHE-RSA-AES256-GCM-SHA384
        minimumProtocolVersion: "1.2"
      useProxyProtocol: false
    logging:
      accessLogFormat: envoy
    metrics:
      address: 0.0.0.0
      port: 8002
    network:
      adminPort: 9001
    service:
      name: envoy
      namespace: projectcontour
  health:
    address: 0.0.0.0
    port: 8000
  httpproxy:
    disablePermitInsecure: false
  leaderElection:
    configmap:
      name: leader-elect
      namespace: projectcontour
    disableLeaderElection: false
    leaseDuration: 15s
    renewDeadline: 10s
    retryPeriod: 2s
  metrics:
    address: 0.0.0.0
    port: 8000
  xdsServer:
    address: 0.0.0.0
    port: 8001
    tls:
      caFile: /certs/ca.crt
      certFile: /certs/tls.crt
      insecure: false
      keyFile: /certs/tls.key
    type: contour
```

The benefit is that we don't have to keep defaults in sync in multiple places.

Reply:
That's how this should work with my last commit. If you specify an empty `spec: {}` you get a full config, otherwise you can override the pieces you'd like and it will still default the rest. 

For example, here's how you'd disable tls communication between contour<>envoy:
```
apiVersion: projectcontour.io/v1alpha1
kind: ContourConfiguration
metadata:
  name: contour
  namespace: projectcontour
spec:
  xdsServer:
    type: contour
    address: 0.0.0.0
    port: 8001
    tls:
      insecure: true
```

```
$ kubectl get contourconfig -o yaml -n projectcontour contour                                                                                                                  

apiVersion: projectcontour.io/v1alpha1
kind: ContourConfiguration
metadata:
  name: contour
  namespace: projectcontour
spec:
  debug:
    kubernetesLogLevel: 0
    logLevel: info
  enableExternalNameService: false
  envoy:
    cluster:
      dnsLookupFamily: auto
    defaultHTTPVersions:
    - http/1.1
    - http/2
    http:
      accessLog: /dev/stdout
      address: 0.0.0.0
      port: 8080
    https:
      accessLog: /dev/stdout
      address: 0.0.0.0
      port: 8443
    listener:
      connectionBalancer: ""
      disableAllowChunkedLength: false
      tls:
        cipherSuites:
        - '[ECDHE-ECDSA-AES128-GCM-SHA256|ECDHE-ECDSA-CHACHA20-POLY1305]'
        - '[ECDHE-RSA-AES128-GCM-SHA256|ECDHE-RSA-CHACHA20-POLY1305]'
        - ECDHE-ECDSA-AES256-GCM-SHA384
        - ECDHE-RSA-AES256-GCM-SHA384
        minimumProtocolVersion: "1.2"
      useProxyProtocol: false
    logging:
      accessLogFormat: envoy
    metrics:
      address: 0.0.0.0
      port: 8002
    network:
      adminPort: 9001
    service:
      name: envoy
      namespace: projectcontour
  health:
    address: 0.0.0.0
    port: 8000
  httpproxy:
    disablePermitInsecure: false
  leaderElection:
    configmap:
      name: leader-elect
      namespace: projectcontour
    disableLeaderElection: false
    leaseDuration: 15s
    renewDeadline: 10s
    retryPeriod: 2s
  metrics:
    address: 0.0.0.0
    port: 8000
  xdsServer:
    address: 0.0.0.0
    port: 8001
    tls:                                # <----- tls bits not defaulted
      insecure: true
    type: contour
```

Reply:
Yeah, I'm just suggesting that we can remove L24 here so we don't have to maintain it/keep it in sync with the lower-level defaults going fwd; even without it, we'll still get all the defaulting when the user specifies `spec: {}`.

Reply:
Ahh I see, removed that. =)